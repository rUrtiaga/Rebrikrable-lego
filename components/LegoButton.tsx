import { Colors } from "@/constants/Colors";
import {
  GestureResponderEvent,
  Pressable,
  StyleSheet,
  Text,
} from "react-native";
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  Easing,
} from "react-native-reanimated";
import Dot, { DotPosition } from "./Dot";

export function LegoButton({
  title,
  onPress,
}: {
  title: string;
  onPress: (event: GestureResponderEvent) => void;
}) {
  const pressAnimation = useSharedValue(0);

  // Animation In function to show that user is pressing button
  const onPressAnimationIn = () => {
    pressAnimation.value = withTiming(6, {
      duration: 130,
      easing: Easing.elastic(0.7),
    });
  };

  // Animation Out function to show that user is unpressing button
  const onPressAnimationOut = () => {
    pressAnimation.value = withTiming(0, {
      duration: 70,
      easing: Easing.linear,
    });
  };

  const animatedStyle = useAnimatedStyle(() => ({
    borderRadius: 8,
    borderTopWidth: pressAnimation.value,
    borderBottomWidth: 5 - pressAnimation.value,
    borderColor: Colors.secondary,
    borderBottomStartRadius: 10,
    shadowColor: "black",
    shadowOffset: { width: 2, height: 5 },
    shadowOpacity: 0.1,
    backgroundColor: Colors.primary,
    height: 50,
    width: 250,
    alignItems: "center",
    justifyContent: "center",
    margin: 10,
  }));

  return (
    <Animated.View style={animatedStyle}>
      <Dot position={DotPosition.topLeft} />
      <Dot position={DotPosition.bottomLeft} />
      <Pressable
        onPress={(p) => {
          onPress(p);
        }}
        onPressIn={onPressAnimationIn}
        onPressOut={onPressAnimationOut}
      >
        <Text style={styles.text}>{title}</Text>
      </Pressable>
      <Dot position={DotPosition.topRight} />
      <Dot position={DotPosition.bottomRight} />
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  text: {
    fontSize: 13,
    fontFamily: "SpaceMono",
    lineHeight: 32,
    color: Colors.text,
  },
});
