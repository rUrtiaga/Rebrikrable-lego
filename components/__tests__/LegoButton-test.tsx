import * as React from 'react';
import renderer from 'react-test-renderer';

import { LegoButton } from '../LegoButton';

it(`Lego-Button renders correctly`, () => {
  const tree = renderer.create(<LegoButton title='this is a title' onPress={()=>{}}/>).toJSON();

  expect(tree).toMatchSnapshot();
});
