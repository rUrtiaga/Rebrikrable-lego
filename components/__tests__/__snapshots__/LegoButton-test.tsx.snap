// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lego-Button renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "alignItems": "center",
          "backgroundColor": "#E5A000",
          "borderBottomStartRadius": 10,
          "borderBottomWidth": 5,
          "borderColor": "#AA7700",
          "borderRadius": 8,
          "borderTopWidth": 0,
          "height": 50,
          "justifyContent": "center",
          "margin": 10,
          "maxWidth": 250,
          "minWidth": 160,
          "shadowColor": "black",
          "shadowOffset": {
            "height": 5,
            "width": 2,
          },
          "shadowOpacity": 0.1,
        },
      }
    }
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#E5A000",
        "borderBottomStartRadius": 10,
        "borderBottomWidth": 5,
        "borderColor": "#AA7700",
        "borderRadius": 8,
        "borderTopWidth": 0,
        "height": 50,
        "justifyContent": "center",
        "margin": 10,
        "maxWidth": 250,
        "minWidth": 160,
        "shadowColor": "black",
        "shadowOffset": {
          "height": 5,
          "width": 2,
        },
        "shadowOpacity": 0.1,
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#FFB200",
          "borderBottomWidth": 1,
          "borderColor": "#AA7700",
          "borderRadius": 50,
          "borderRightWidth": 1,
          "elevation": 4,
          "height": 14,
          "left": 4,
          "position": "absolute",
          "shadowColor": "black",
          "shadowOffset": {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.3,
          "top": 4,
          "width": 14,
          "zIndex": 50,
        }
      }
    />
    <View
      style={
        {
          "backgroundColor": "#FFB200",
          "borderBottomWidth": 1,
          "borderColor": "#AA7700",
          "borderRadius": 50,
          "borderRightWidth": 1,
          "bottom": 4,
          "elevation": 4,
          "height": 14,
          "left": 4,
          "position": "absolute",
          "shadowColor": "black",
          "shadowOffset": {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.3,
          "width": 14,
          "zIndex": 50,
        }
      }
    />
    <Text
      style={
        {
          "color": "white",
          "fontFamily": "SpaceMono",
          "fontSize": 13,
          "lineHeight": 32,
        }
      }
    >
      this is a title
    </Text>
    <View
      style={
        {
          "backgroundColor": "#FFB200",
          "borderBottomWidth": 1,
          "borderColor": "#AA7700",
          "borderRadius": 50,
          "borderRightWidth": 1,
          "elevation": 4,
          "height": 14,
          "position": "absolute",
          "right": 4,
          "shadowColor": "black",
          "shadowOffset": {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.3,
          "top": 4,
          "width": 14,
          "zIndex": 50,
        }
      }
    />
    <View
      style={
        {
          "backgroundColor": "#FFB200",
          "borderBottomWidth": 1,
          "borderColor": "#AA7700",
          "borderRadius": 50,
          "borderRightWidth": 1,
          "bottom": 4,
          "elevation": 4,
          "height": 14,
          "position": "absolute",
          "right": 4,
          "shadowColor": "black",
          "shadowOffset": {
            "height": 1,
            "width": 1,
          },
          "shadowOpacity": 0.3,
          "width": 14,
          "zIndex": 50,
        }
      }
    />
  </View>
</View>
`;
